File: formcraft-plm/src/main/java/fr/formcraft/model/enums/NcStatus.java
package fr.formcraft.model.enums;

/** Lifecycle state of a non-conformance record. */
public enum NcStatus {
    OPEN,
    IN_PROGRESS,
    CLOSED;

    public boolean canTransitionTo(NcStatus target) {
        return switch (this) {
            case OPEN -> target == IN_PROGRESS;
            case IN_PROGRESS -> target == CLOSED || target == OPEN;
            case CLOSED -> false;
        };
    }
}

File: app/enums/nc_status.py
import enum


class NcStatus(str, enum.Enum):
    OPEN = "OPEN"
    IN_PROGRESS = "IN_PROGRESS"
    CLOSED = "CLOSED"

    def can_transition_to(self, target: "NcStatus") -> bool:
        match self:
            case NcStatus.OPEN:
                return target == NcStatus.IN_PROGRESS
            case NcStatus.IN_PROGRESS:
                return target in (NcStatus.CLOSED, NcStatus.OPEN)
            case NcStatus.CLOSED:
                return False
